File: frontend/packages/api-client/webhooks.ts
import { Axios } from 'axios';
import format from 'string-format';

import {
  WebhookCreationRequestInput,
  Webhook,
  QueryFilter,
  QueryFilteredResult,
  APIResponse,
} from '@dinnerdonebetter/models';

import { backendRoutes } from './routes';

export async function createWebhook(client: Axios, input: WebhookCreationRequestInput): Promise<Webhook> {
  return new Promise(async function (resolve, reject) {
    const response = await client.post<APIResponse<Webhook>>(backendRoutes.VALID_PREPARATIONS, input);

    if (response.data.error) {
      reject(new Error(response.data.error.message));
    }

    resolve(response.data.data);
  });
}

export async function getWebhook(client: Axios, WebhookID: string): Promise<Webhook> {
  return new Promise(async function (resolve, reject) {
    const response = await client.get<APIResponse<Webhook>>(format(backendRoutes.VALID_PREPARATION, WebhookID));

    if (response.data.error) {
      reject(new Error(response.data.error.message));
    }

    resolve(response.data.data);
  });
}

export async function getWebhooks(
  client: Axios,
  filter: QueryFilter = QueryFilter.Default(),
): Promise<QueryFilteredResult<Webhook>> {
  return new Promise(async function (resolve, reject) {
    const response = await client.get<APIResponse<Webhook[]>>(backendRoutes.VALID_PREPARATIONS, {
      params: filter.asRecord(),
    });

    if (response.data.error) {
      reject(new Error(response.data.error.message));
    }

    const result = new QueryFilteredResult<Webhook>({
      data: response.data.data,
      totalCount: response.data.pagination?.totalCount,
      page: response.data.pagination?.page,
      filteredCount: response.data.pagination?.filteredCount,
    });

    resolve(result);
  });
}

export async function deleteWebhook(client: Axios, WebhookID: string): Promise<Webhook> {
  return new Promise(async function (resolve, reject) {
    const response = await client.delete<APIResponse<Webhook>>(format(backendRoutes.VALID_PREPARATION, WebhookID));

    if (response.data.error) {
      reject(new Error(response.data.error.message));
    }

    resolve(response.data.data);
  });
}

export async function searchForWebhooks(client: Axios, query: string): Promise<Webhook[]> {
  return new Promise(async function (resolve, reject) {
    const searchURL = `${backendRoutes.VALID_PREPARATIONS_SEARCH}?q=${encodeURIComponent(query)}`;
    const response = await client.get<APIResponse<Webhook[]>>(searchURL);

    if (response.data.error) {
      reject(new Error(response.data.error.message));
    }

    resolve(response.data.data);
  });
}

File: backend/pkg/apiclient/webhooks.go
package apiclient

import (
	"context"

	"github.com/dinnerdonebetter/backend/internal/observability"
	"github.com/dinnerdonebetter/backend/internal/observability/keys"
	"github.com/dinnerdonebetter/backend/internal/observability/tracing"
	"github.com/dinnerdonebetter/backend/pkg/apiclient/generated"
	"github.com/dinnerdonebetter/backend/pkg/types"
)

// GetWebhook retrieves a webhook.
func (c *Client) GetWebhook(ctx context.Context, webhookID string) (*types.Webhook, error) {
	ctx, span := c.tracer.StartSpan(ctx)
	defer span.End()

	logger := c.logger.Clone()

	if webhookID == "" {
		return nil, ErrInvalidIDProvided
	}
	logger = logger.WithValue(keys.WebhookIDKey, webhookID)

	res, err := c.authedGeneratedClient.GetWebhook(ctx, webhookID)
	if err != nil {
		return nil, observability.PrepareAndLogError(err, logger, span, "getting webhook")
	}
	defer c.closeResponseBody(ctx, res)

	var apiResponse *types.APIResponse[*types.Webhook]
	if err = c.unmarshalBody(ctx, res, &apiResponse); err != nil {
		return nil, observability.PrepareAndLogError(err, logger, span, "loading webhook response")
	}

	if err = apiResponse.Error.AsError(); err != nil {
		return nil, err
	}

	return apiResponse.Data, nil
}

// GetWebhooks gets a list of webhooks.
func (c *Client) GetWebhooks(ctx context.Context, filter *types.QueryFilter) (*types.QueryFilteredResult[types.Webhook], error) {
	ctx, span := c.tracer.StartSpan(ctx)
	defer span.End()

	if filter == nil {
		filter = types.DefaultQueryFilter()
	}

	logger := c.logger.Clone()
	logger = filter.AttachToLogger(logger)
	tracing.AttachQueryFilterToSpan(span, filter)

	params := &generated.GetWebhooksParams{}
	c.copyType(params, filter)

	res, err := c.authedGeneratedClient.GetWebhooks(ctx, params)
	if err != nil {
		return nil, observability.PrepareAndLogError(err, logger, span, "getting webhooks")
	}
	defer c.closeResponseBody(ctx, res)

	var apiResponse *types.APIResponse[[]*types.Webhook]
	if err = c.unmarshalBody(ctx, res, &apiResponse); err != nil {
		return nil, observability.PrepareAndLogError(err, logger, span, "loading webhooks response")
	}

	if err = apiResponse.Error.AsError(); err != nil {
		return nil, err
	}

	webhooks := &types.QueryFilteredResult[types.Webhook]{
		Data:       apiResponse.Data,
		Pagination: *apiResponse.Pagination,
	}

	return webhooks, nil
}

// CreateWebhook creates a webhook.
func (c *Client) CreateWebhook(ctx context.Context, input *types.WebhookCreationRequestInput) (*types.Webhook, error) {
	ctx, span := c.tracer.StartSpan(ctx)
	defer span.End()

	logger := c.logger.Clone()

	if input == nil {
		return nil, ErrNilInputProvided
	}

	if err := input.ValidateWithContext(ctx); err != nil {
		return nil, observability.PrepareAndLogError(err, logger, span, "validating input")
	}

	body := generated.CreateWebhookJSONRequestBody{}
	c.copyType(&body, input)

	res, err := c.authedGeneratedClient.CreateWebhook(ctx, body)
	if err != nil {
		return nil, observability.PrepareAndLogError(err, logger, span, "creating webhook")
	}
	defer c.closeResponseBody(ctx, res)

	var apiResponse *types.APIResponse[*types.Webhook]
	if err = c.unmarshalBody(ctx, res, &apiResponse); err != nil {
		return nil, observability.PrepareAndLogError(err, logger, span, "parsing webhook response")
	}

	if err = apiResponse.Error.AsError(); err != nil {
		return nil, err
	}

	return apiResponse.Data, nil
}

// ArchiveWebhook archives a webhook.
func (c *Client) ArchiveWebhook(ctx context.Context, webhookID string) error {
	ctx, span := c.tracer.StartSpan(ctx)
	defer span.End()

	logger := c.logger.Clone()

	if webhookID == "" {
		return ErrInvalidIDProvided
	}
	logger = logger.WithValue(keys.WebhookIDKey, webhookID)

	res, err := c.authedGeneratedClient.ArchiveWebhook(ctx, webhookID)
	if err != nil {
		return observability.PrepareAndLogError(err, logger, span, "archiving webhook")
	}
	defer c.closeResponseBody(ctx, res)

	var apiResponse *types.APIResponse[*types.Webhook]
	if err = c.unmarshalBody(ctx, res, &apiResponse); err != nil {
		return observability.PrepareAndLogError(err, logger, span, "loading archiving webhook response")
	}

	if err = apiResponse.Error.AsError(); err != nil {
		return err
	}

	return nil
}

// AddWebhookTriggerEvent adds a webhook trigger event.
func (c *Client) AddWebhookTriggerEvent(ctx context.Context, webhookID string, input *types.WebhookTriggerEventCreationRequestInput) (*types.WebhookTriggerEvent, error) {
	ctx, span := c.tracer.StartSpan(ctx)
	defer span.End()

	logger := c.logger.Clone()

	if webhookID == "" {
		return nil, ErrInvalidIDProvided
	}
	logger = logger.WithValue(keys.WebhookIDKey, webhookID)

	if input == nil {
		return nil, ErrNilInputProvided
	}

	if err := input.ValidateWithContext(ctx); err != nil {
		return nil, observability.PrepareAndLogError(err, logger, span, "validating input")
	}

	body := generated.CreateWebhookTriggerEventJSONRequestBody{}
	c.copyType(&body, input)

	res, err := c.authedGeneratedClient.CreateWebhookTriggerEvent(ctx, webhookID, body)
	if err != nil {
		return nil, observability.PrepareAndLogError(err, logger, span, "creating webhook trigger event")
	}
	defer c.closeResponseBody(ctx, res)

	var apiResponse *types.APIResponse[*types.WebhookTriggerEvent]
	if err = c.unmarshalBody(ctx, res, &apiResponse); err != nil {
		return nil, observability.PrepareAndLogError(err, logger, span, "loading webhook trigger event creation response")
	}

	if err = apiResponse.Error.AsError(); err != nil {
		return nil, err
	}

	return apiResponse.Data, nil
}

// ArchiveWebhookTriggerEvent archives a webhook trigger event.
func (c *Client) ArchiveWebhookTriggerEvent(ctx context.Context, webhookID, webhookTriggerEventID string) error {
	ctx, span := c.tracer.StartSpan(ctx)
	defer span.End()

	logger := c.logger.Clone()

	if webhookID == "" {
		return ErrInvalidIDProvided
	}
	logger = logger.WithValue(keys.WebhookIDKey, webhookID)

	if webhookTriggerEventID == "" {
		return ErrInvalidIDProvided
	}
	logger = logger.WithValue(keys.WebhookTriggerEventIDKey, webhookTriggerEventID)

	res, err := c.authedGeneratedClient.ArchiveWebhookTriggerEvent(ctx, webhookID, webhookTriggerEventID)
	if err != nil {
		return observability.PrepareAndLogError(err, logger, span, "archiving webhook trigger event")
	}
	defer c.closeResponseBody(ctx, res)

	var apiResponse *types.APIResponse[*types.Webhook]
	if err = c.unmarshalBody(ctx, res, &apiResponse); err != nil {
		return observability.PrepareAndLogError(err, logger, span, "loading archiving webhook trigger event response")
	}

	if err = apiResponse.Error.AsError(); err != nil {
		return err
	}

	return nil
}
